torch/jit.py

```diff
@@ -1,26 +1,375 @@
+    def _trace(inner):
+        return TracedModule(inner, nderivs=nderivs, params=params)
+    if callable(arg):
+        return _trace(arg)
+    else:
+        return _trace
+
+
+# TODO: Formulating it this way means that state_dict of the traced thing
+# looks different, etc (because there's extra nesting).  Figure out a
+# way to avoid this
+class TracedModule(Module):
+    def __init__(self, inner, params=tuple(), nderivs=0):
+        super(TracedModule, self).__init__()
+        # inner may be a Module, or it may be an arbitrary callable
+        self.inner = inner
+        self.params = ParameterList(list(params))
+        self.nderivs = nderivs
+
+    def forward(self, *args):
+        # TODO: Possible optimization: use the unflattened
+        # output so we don't unflatten it when we get out
+        # NB: Not a method because trace_func_raw can't deal
+        # with methods
+        @raw_trace(nderivs=self.nderivs)
+        def traced_inner(in_vars, in_struct):
+            # NB: Uncommenting this line should be equivalent
+            # args, _params = _unflatten(in_vars, in_struct)
+            return _flatten(self.inner(*args))
+
+        in_vars, in_struct = _flatten(args, self.state_dict(keep_vars=True).values())
+        trace, (out_vars, out_struct) = traced_inner(in_vars, in_struct)
+        out, extra = _unflatten(out_vars, out_struct)
+        assert len(extra) == 0
+        return trace, out
+
+
+# Functional version that assumes that all parameters are explicitly
+# specified
+def raw_trace(nderivs=0):
+    def _raw_trace(f):
+        @functools.wraps(f)
+        def wrapper(in_vars, in_struct=None):
+            trace = torch._C._tracer_enter(in_vars, nderivs)
+            out_vars, out_struct = f(in_vars, in_struct)
+            torch._C._tracer_exit(out_vars)
+            return trace, (out_vars, out_struct)
+        return wrapper
+    return _raw_trace
+
+
+# Lifecycle of a CompiledModule:
+# - It is given an underlying function, which knows how to actually
+#   execute the code that we want to compile.
+# - When we encounter an input configuration for which we don't
+#   have an optimized trace, we run the underlying function, tracing its
+#   result.  The trace is not done yet, so we save it into our set of pending
+#   traces for that configuration.
+# - When we encounter an input configuration whose trace is "ready"
+#   (that is, we've seen all of the passes, so the trace contains
+#   forwards/backwards/etc), we compile it, and then register this
+#   as the compiled trace.
+# - When we encounter an input configuration whose trace is compiled,
+#   we just directly run the compiled trace.
+class CompiledModule(Module):
+    def __init__(self, inner, params=tuple(), **kwargs):
+        super(CompiledModule, self).__init__()
+        self.inner = inner
+        self.params = ParameterList(list(params))
+        self.kwargs = kwargs
+        self.ktrace_cache = {}
```

Comment:
why the k prefix?

Reply:
To distinguish between a trace (TracingState) and a TraceForKey (ktrace)